### PR TITLE
Avoid some `InferCtxt::build` calls

### PR DESCRIPTION
**Pull request by compiler-errors**
Either because we're inside of an `InferCtxt` already, or because we're not in a place where we'd ever see inference vars.

r? types

### CODE REVIEW

File: src/test/ui/return/tail-expr-as-potential-return.rs [hunk 1]
```diff
@@ -30,3 +29,19 @@ async fn bar(x: bool) -> Result<f64, i32> {
     }
     Ok(42.0)
 }
+
```
Comment (jackh726) [inline src/test/ui/return/tail-expr-as-potential-return.rs:32]: Why the change to this test?

Comment (compiler-errors) [reply src/test/ui/return/tail-expr-as-potential-return.rs:32]: i moved a normalize call so that the suggestion now works when theres a projection in the return type of an async fn. before this commit, it was missing. 

sorry, should've commented this drive-by.

File: compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs [hunk 2]
```diff
@@ -921,22 +921,19 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                    .unwrap_or_else(|| {
+                        span_bug!(
+                            fn_decl.output.span(),
+                            "failed to get output type of async function"
+                        )
+                    })
+                    .skip_binder(),
                 hir::IsAsync::NotAsync => ty,
             };
+            let ty = self.normalize_associated_types_in(expr.span, ty);
```
Comment (compiler-errors) [inline compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs:936]: For the record, moving this normalize call is what allows the UI test I committed to properly emit:

help: you might have meant to return this value
   |
LL |         return 1i32;
   |         ++++++     +

**Comment by jackh726** [comment]
r=me

**Comment by compiler-errors** [comment]
The queue is short and the failure looks unrelated...?

@bors retry
